src/ynEIGEN/TokenStakingNode.sol

```diff
@@ -160,81 +155,92 @@ contract TokenStakingNode is
-                _withdrawals,
-                _tokens,
-                _middlewareTimesIndexes,
-                _receiveAsTokens
-            );
-        }
+        _delegationManager.completeQueuedWithdrawal(
+            withdrawal,
+            _tokens,
+            middlewareTimesIndex,
+            true // receiveAsTokens
+        );
 
         uint256 _actualAmountOut = _token.balanceOf(address(this)) - _balanceBefore;
         IWrapper _wrapper = IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).wrapper();
         IERC20(_token).forceApprove(address(_wrapper), _actualAmountOut); // NOTE: approving also token that will not be transferred
         (_actualAmountOut, _token) = _wrapper.wrap(_actualAmountOut, _token);
-
+        uint256 _shares = withdrawal.shares[0];
         queuedShares[_strategy] -= _shares;
         withdrawn[_token] += _actualAmountOut;
 
         if (updateTokenStakingNodesBalances) {
             // Actual balance changes only if slashing occured. choose to update here
             // only if the off-chain considers it necessary to save gas
-            IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager()).updateTokenStakingNodesBalances(
-                _token
-            );
+            IYieldNestStrategyManager(tokenStakingNodesManager.yieldNestStrategyManager())
+                .updateTokenStakingNodesBalances(_token);
         }
-
         emit CompletedQueuedWithdrawals(_shares, _actualAmountOut, address(_strategy));
     }
 
+    /**
+     * @notice Struct containing strategy and shares information
+     * @param strategy The strategy contract address
+     * @param shares The number of shares
+     */
+    struct StrategyShares {
+        IStrategy strategy;
+        uint256 shares;
+    }
+
+    /**
+     * @notice Completes queued withdrawals with receiveAsTokens set to false
+     * @param withdrawals Array of withdrawals to complete
+     * @param middlewareTimesIndexes Array of middleware times indexes
+     * @param strategies Array of strategies to withdraw from
+     */
+    function completeQueuedWithdrawalsAsShares(
+        IDelegationManager.Withdrawal[] calldata withdrawals,
+        uint256[] calldata middlewareTimesIndexes,
+        IStrategy[] calldata strategies
+    ) external onlyDelegator onlyWhenSynchronized {
+        if (withdrawals.length != middlewareTimesIndexes.length || withdrawals.length != strategies.length) {
+            revert ArrayLengthMismatch();
+        }
+
+        IDelegationManager delegationManager = IDelegationManager(address(tokenStakingNodesManager.delegationManager()));
+        IERC20[][] memory _tokens = new IERC20[][](withdrawals.length);
+        bool[] memory _receiveAsTokens = new bool[](withdrawals.length);
+        // Decrease queued shares for each strategy
+        for (uint256 i = 0; i < withdrawals.length; i++) {
+            queuedShares[strategies[i]] -= withdrawals[i].shares[0];
```

Comment:
this is potentially submoptimal gas-wise if we use for many withdrawals for the same asset.

That's what the initial version of the code was doing by counting totals per asset.

However given that completeQueuedWithdrawalsAsShares only has one use case now which is to recover from operator undelegation and infrequent use, this is fine